### PR TITLE
Intel GPU Runtime Upstreaming for Event

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118613
* #118528
* #118523
* #118091
* __->__ #117734
* #117619
* #117611

# Motivation
As mentioned in [[RFC] Intel GPU Runtime Upstreaming](https://github.com/pytorch/pytorch/issues/114842), the next runtime component we would like to upstream is `Event` which handles the status of an operation that is being executed. Typically, in some circumstances, we can fine-grain control of the operation execution via `Event`.

# Design
`XPUEvent` is a movable but not a copyable wrapper around sycl event. It should be created lazily on an XPU device when recording an `XPUStream`. Meanwhile, `XPUEvent` can wait for another `XPUEvent` or all the submitted kernels on an `XPUStream` to complete. Align to the other backend, the C++ files related to `Event` will be placed in `aten/src/ATen/xpu` folder. For frontend code, `XPUEvent` runtime API will be bound to Python `torch.xpu.Event`. The corresponding C++ code will be placed in `torch/csrc/xpu/Event.cpp` and Python code will be placed in `torch/xpu/streams.py` respectively.

# Additional Context
It is worth mentioning that the `elapsed_time` method is temporarily not supported by `XPUEvent`. We will be adding support for it soon. Meanwhile `XPUEvent` doesn't support IPC from different processes. For the other parts, we have almost a 1:1 mapping with CUDA.

lack of the below APIs:
- `torch.cuda.Event.ipc_handle`
- `CUDAEvent`'s constructor with `IpcEventHandle`

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by EikanWang — APPROVED** [review]
LGTM. Some minor comments.

File: aten/src/ATen/xpu/XPUEvent.h [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+  ~XPUEvent() {}
+
+  XPUEvent(const XPUEvent&) = delete;
+  XPUEvent& operator=(const XPUEvent&) = delete;
+
+  XPUEvent(XPUEvent&& other) noexcept {
+    moveHelper(std::move(other));
+  }
+  XPUEvent& operator=(XPUEvent&& other) noexcept {
```
Comment (gujinghui) [inline aten/src/ATen/xpu/XPUEvent.h:26]: Pls. add a blank line btw two functions.

File: aten/src/ATen/xpu/XPUEvent.h [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+      return at::Device(at::kXPU, device_index_);
+    } else {
+      return {};
+    }
+  }
+
+  bool isCreated() const {
+    return (event_.get() != nullptr);
+  }
+  DeviceIndex device_index() const {
```
Comment (gujinghui) [inline aten/src/ATen/xpu/XPUEvent.h:48]: ditto.

File: aten/src/ATen/xpu/XPUEvent.h [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+ * XPUEvent are movable not copyable wrappers around SYCL event. XPUEvent are
+ * constructed lazily when first recorded. It has a device, and this device is
+ * acquired from the first recording stream. Later streams that record the event
+ * must match the same device.
+ */
+struct TORCH_XPU_API XPUEvent {
+  // Constructors
+  XPUEvent() noexcept = default;
+  XPUEvent(bool enable_timing) noexcept : enable_timing_{enable_timing} {}
+
```
Comment (gujinghui) [inline aten/src/ATen/xpu/XPUEvent.h:17]: The XPUEvent is incomplete, because of lack of device guard?

File: aten/src/ATen/xpu/XPUEvent.h [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    return *event_;
+  }
+
+  bool query() const {
+    using namespace sycl::info;
+    if (!isCreated()) {
+      return true;
+    }
+
+    return event().get_info<event::command_execution_status>() ==
```
Comment (gujinghui) [inline aten/src/ATen/xpu/XPUEvent.h:61]: Is this on par with cuda event, cudaEventQuery?

File: aten/src/ATen/xpu/XPUEvent.h [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        event_command_status::complete;
+  }
+
+  void record() {
+    record(getCurrentXPUStream());
+  }
+
+  void recordOnce(const XPUStream& stream) {
+    if (!isCreated()) {
```
Comment (gujinghui) [inline aten/src/ATen/xpu/XPUEvent.h:70]: cuda is using was_recorded_, why we check isCreated here?

File: aten/src/ATen/xpu/XPUEvent.h [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+      TORCH_CHECK(
+          device_index_ == stream.device_index(),
+          "Event device ",
+          device_index_,
+          " does not match recording stream's device ",
+          stream.device_index(),
+          ".");
+      event_.reset();
+    }
+    event_ = std::make_unique<sycl::event>(
```
Comment (gujinghui) [inline aten/src/ATen/xpu/XPUEvent.h:88]: why we do not have GPUTrace? It should be device-agnostic?

File: aten/src/ATen/xpu/XPUEvent.h [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        stream.queue().ext_oneapi_submit_barrier());
+  }
+
+  void block(const XPUStream& stream) {
+    if (isCreated()) {
+      std::vector<sycl::event> event_list{event()};
+      // Make this stream wait until event_ is completed.
+      stream.queue().ext_oneapi_submit_barrier(event_list);
```
Comment (gujinghui) [inline aten/src/ATen/xpu/XPUEvent.h:96]: Confused me. block means wait on host. Why we submit a new barrier here?

File: aten/src/ATen/xpu/XPUEvent.h [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        "Both events must be created with argument 'enable_timing=True'.");
+    // TODO: provides the ability to time the execution of commands in a SYCL
+    // queue without enabling profiling on the entire queue
+    TORCH_CHECK_NOT_IMPLEMENTED(
+        false, "elapsed_time is not supported by XPUEvent.");
+  }
+
+  void synchronize() const {
+    if (isCreated()) {
+      event().wait_and_throw();
```
Comment (gujinghui) [inline aten/src/ATen/xpu/XPUEvent.h:118]: Here we should sync with event_ member, instead of a new empty event?

File: aten/src/ATen/xpu/XPUEvent.h [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    }
+  }
+
+ private:
+  bool enable_timing_ = false;
+  DeviceIndex device_index_ = -1;
+  // Only need to track the last event, as events in an in-order queue are
+  // executed sequentially.
+  std::unique_ptr<sycl::event> event_;
```
Comment (gujinghui) [inline aten/src/ATen/xpu/XPUEvent.h:127]: event list is a wrong/legacy impl.
we should follow cuda, single event is enough.

File: aten/src/ATen/xpu/XPUEvent.h [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+  optional<at::Device> device() const {
+    if (isCreated()) {
+      return at::Device(at::kXPU, device_index_);
+    } else {
+      return {};
+    }
+  }
+
+  bool isCreated() const {
```
[the remark this answers is not in the document]
Comment (guangyey) [reply aten/src/ATen/xpu/XPUEvent.h:45]: done. rename to `is_created`

File: aten/src/ATen/xpu/XPUEvent.h [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    return (event_.get() != nullptr);
+  }
+  DeviceIndex device_index() const {
+    return device_index_;
+  }
+  sycl::event& event() const {
+    return *event_;
+  }
```
[the remark this answers is not in the document]
Comment (guangyey) [reply aten/src/ATen/xpu/XPUEvent.h:53]: done.

File: aten/src/ATen/xpu/XPUEvent.h (hunk 1 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUEvent.h:26]: done.

File: aten/src/ATen/xpu/XPUEvent.h (hunk 3 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUEvent.h:17]: device guard is not necessary. XPUEvent currently does not support `IPC`.

File: aten/src/ATen/xpu/XPUEvent.h (hunk 4 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUEvent.h:61]: yes, align to `cudaEventQuery`.

File: aten/src/ATen/xpu/XPUEvent.h (hunk 5 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUEvent.h:70]: There are two methods to create a `CUDAEvent`: one is `record`, and the other one is `ipc_handle`. So for cuda, `was_recorded_` is not equivalent to `isCreated`. `XPUEvent` is created only by `record`. So `was_recorded` is redundant to `XPUEvent`.

File: aten/src/ATen/xpu/XPUEvent.h (hunk 6 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUEvent.h:88]: `GPUTrace` should be device-agnostic. Currently, it has some cuda-specific code.

File: aten/src/ATen/xpu/XPUEvent.h (hunk 3 above)
Comment (gujinghui) [reply aten/src/ATen/xpu/XPUEvent.h:17]: SYCL already has IPC API now?

Comment (guangyey) [reply aten/src/ATen/xpu/XPUEvent.h:17]: No. SYCL doesn't have, L0 has.

**Review by gujinghui — APPROVED** [review]

File: aten/src/ATen/xpu/XPUEvent.h (hunk 3 above)
Comment (jgong5) [reply aten/src/ATen/xpu/XPUEvent.h:17]: What's our plan of supporting the IPC event? If we do not support it right now, suggest to add a note in the code explaining the reasons and impact.

File: aten/src/ATen/xpu/XPUEvent.h (hunk 6 above)
Comment (jgong5) [reply aten/src/ATen/xpu/XPUEvent.h:91]: Can you elaborate? I didn't find any cuda-specific code. If it does, please add a TODO here.

File: aten/src/ATen/xpu/XPUEvent.h
Comment (jgong5) [inline aten/src/ATen/xpu/XPUEvent.h:1]: To facilitate the reviews, I would suggest we highlight the difference from the CUDA counterparts in the PR description or as review comments in the future.

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/xpu/XPUEvent.h (hunk 3 above)
Comment (gujinghui) [reply aten/src/ATen/xpu/XPUEvent.h:17]: Yes. We should add note for it. @guangyey

File: aten/src/ATen/xpu/XPUEvent.h (hunk 6 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUEvent.h:91]: `GPUTrace` is CUDA-bias, curently. see https://github.com/pytorch/pytorch/blob/995f69623d6bee50f0d413c546c838e1df1c4dcd/torch/csrc/PyInterpreter.cpp#L92-L95
and there is a hook `torch.utils._cuda_trace` in `CONCRETE_TRACE_CUDA`, see 
https://github.com/pytorch/pytorch/blob/995f69623d6bee50f0d413c546c838e1df1c4dcd/torch/csrc/PyInterpreter.cpp#L21-L30.

File: aten/src/ATen/xpu/XPUEvent.h
Comment (guangyey) [reply aten/src/ATen/xpu/XPUEvent.h:1]: OK, thanks.

File: aten/src/ATen/xpu/XPUEvent.h [hunk 12]
```diff
@@ -9,6 +9,12 @@ namespace at::xpu {
  * constructed lazily when first recorded. It has a device, and this device is
  * acquired from the first recording stream. Later streams that record the event
  * must match the same device.
+ *
+ * Currently, XPUEvent does NOT support to export an inter-process event from
+ * another process via inter-process comunication(IPC). So it means that
+ * inter-process communication for event handles between different processes is
+ * not available. This could impact some applications that rely on cross-process
+ * synchronization and communication.
```
Comment (jgong5) [inline aten/src/ATen/xpu/XPUEvent.h:17]: Is this a TODO for future? If so, please add a TODO tag here.

**Review by jgong5 — APPROVED** [review]

**Comment by jgong5** [comment]
Thanks for adding additional context in the PR description. Please make sure all the TODOs are marked in the code as well.

File: test/test_xpu.py [hunk 13]
```diff
@@ -92,6 +92,25 @@ def test_stream_priority(self):
         self.assertEqual(high, s1.priority)
         self.assertEqual(torch.device("xpu:0"), s1.device)
 
+    def test_stream_event_repr(self):
+        s = torch.xpu.current_stream()
+        self.assertTrue("torch.xpu.Stream" in s.__repr__())
```
Comment (albanD) [inline test/test_xpu.py:97]: nit: usually str(s) is used

File: test/test_xpu.py [hunk 14]
```diff
@@ -92,6 +92,25 @@ def test_stream_priority(self):
...
+        self.assertTrue("torch.xpu.Event" in e.__repr__())
+
+    def test_events(self):
+        stream = torch.xpu.current_stream()
+        event = torch.xpu.Event()
+        self.assertTrue(event.query())
+        start_event = torch.xpu.Event()
+        stream.record_event(start_event)
+        stream.record_event(event)
+        self.assertFalse(event.query())
```
Comment (albanD) [inline test/test_xpu.py:110]: Isn't this racy? It's unlikely but the event might actually be completed already here?

File: torch/csrc/xpu/Event.cpp [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+  THXPEvent* self = (THXPEvent*)ptr.get();
+
+  new (&self->xpu_event) at::xpu::XPUEvent(enable_timing);
+
+  return (PyObject*)ptr.release();
+  END_HANDLE_TH_ERRORS
+}
+
+static void THXPEvent_dealloc(THXPEvent* self) {
+  self->xpu_event.~XPUEvent();
```
Comment (albanD) [inline torch/csrc/xpu/Event.cpp:43]: FYI this now has a gil release in the cuda side https://github.com/pytorch/pytorch/pull/119393
If the event destruction might trigger non-trivial code and block, you should do the same

File: torch/csrc/xpu/Event.cpp [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+  }
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays,
+// cppcoreguidelines-avoid-non-const-global-variables, modernize-avoid-c-arrays)
+static struct PyGetSetDef THXPEvent_properties[] = {
+    {"device", (getter)THXPEvent_get_device, nullptr, nullptr, nullptr},
+    {"sycl_event", (getter)THXPEvent_get_sycl_event, nullptr, nullptr, nullptr},
```
Comment (albanD) [inline torch/csrc/xpu/Event.cpp:110]: Could we name this something more generic in preparation for more uniformity?
Like `raw_event_ptr`. This will be one less public API we have to change going forward.

**Review by albanD — COMMENTED** [review]
I think the shared files are the same as the ones in the Stream PR.

File: test/test_xpu.py (hunk 13 above)
Comment (guangyey) [reply test/test_xpu.py]: Done.

File: test/test_xpu.py (hunk 14 above)
Comment (guangyey) [reply test/test_xpu.py]: Yes, you are right. Remove this assertation.

File: torch/csrc/xpu/Event.cpp (hunk 15 above)
Comment (guangyey) [reply torch/csrc/xpu/Event.cpp:43]: Thanks for your information. If `cudaEventDestroy` is called while the associated event hasn't been completed, the CUDA runtime will wait for those operations to complete asynchronously before destroying the CUDA event. I think SYCL runtime might also encounter similar issues, although I cannot confirm it one hundred percent.

File: torch/csrc/xpu/Event.cpp (hunk 16 above)
Comment (guangyey) [reply torch/csrc/xpu/Event.cpp:113]: `raw_event_ptr` is OK. But currently, we would like to align with CUDA. For `torch.cuda.Event`, the counterpart names it as `cuda_event`, also as a raw pointer.

File: torch/csrc/xpu/Event.cpp (hunk 15 above)
Comment (guangyey) [reply torch/csrc/xpu/Event.cpp]: Add gil release code.

**Comment by guangyey** [comment]
> I think the shared files are the same as the ones in the Stream PR.

I put some shared code in `c10/xpu/test/impl/XPUTest.h` to reuse them in different UTs. Do I capture your point?

**Comment by guangyey** [comment]
@albanD  May I know if we captured your point and addressed comments?

File: torch/csrc/xpu/Event.cpp (hunk 16 above)
Comment (albanD) [reply torch/csrc/xpu/Event.cpp:113]: I don't think aligning with cuda gives any benefit here? This is an internal field that is only used by the python class you write.

File: aten/src/ATen/xpu/XPUEvent.h [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+ */
+struct TORCH_XPU_API XPUEvent {
+  // Constructors
+  XPUEvent() noexcept = default;
+  XPUEvent(bool enable_timing) noexcept : enable_timing_{enable_timing} {}
+
+  ~XPUEvent() {}
```
Comment (malfet) [inline aten/src/ATen/xpu/XPUEvent.h:24]: Nit
```suggestion
  ~XPUEvent() = default;
```

File: aten/src/ATen/xpu/XPUEvent.h [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+ * Currently, XPUEvent does NOT support to export an inter-process event from
+ * another process via inter-process comunication(IPC). So it means that
+ * inter-process communication for event handles between different processes is
+ * not available. This could impact some applications that rely on cross-process
+ * synchronization and communication.
+ */
+struct TORCH_XPU_API XPUEvent {
+  // Constructors
+  XPUEvent() noexcept = default;
+  XPUEvent(bool enable_timing) noexcept : enable_timing_{enable_timing} {}
```
Comment (malfet) [inline aten/src/ATen/xpu/XPUEvent.h:22]: Is there a benefit of having 2 constructors instead of one with default argument?
```suggestion
  XPUEvent(bool enable_timing = false) noexcept : enable_timing_{enable_timing} {}
```

File: aten/src/ATen/xpu/XPUEvent.h [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+  DeviceIndex device_index_ = -1;
+  // Only need to track the last event, as events in an in-order queue are
+  // executed sequentially.
+  std::unique_ptr<sycl::event> event_;
+
+  void moveHelper(XPUEvent&& other) {
+    std::swap(enable_timing_, other.enable_timing_);
+    std::swap(device_index_, other.device_index_);
+    std::swap(event_, other.event_);
+  }
```
Comment (malfet) [inline aten/src/ATen/xpu/XPUEvent.h:142]: Why is this necessary? Isn't it a default move constructor behavior?

File: aten/src/ATen/xpu/XPUEvent.h [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+  optional<at::Device> device() const {
+    if (isCreated()) {
+      return at::Device(at::kXPU, device_index_);
+    } else {
+      return {};
+    }
+  }
+
+  bool isCreated() const {
+    return (event_.get() != nullptr);
```
Comment (malfet) [inline aten/src/ATen/xpu/XPUEvent.h:53]: Nit (replies on https://en.cppreference.com/w/cpp/memory/unique_ptr/operator_bool )
```suggestion
    return event_;
```

File: aten/src/ATen/xpu/XPUEvent.h [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+
+  optional<at::Device> device() const {
+    if (isCreated()) {
+      return at::Device(at::kXPU, device_index_);
+    } else {
+      return {};
+    }
+  }
+
+  bool isCreated() const {
```
Comment (malfet) [inline aten/src/ATen/xpu/XPUEvent.h:52]: Nit
```suggestion
  inline bool isCreated() const {
```

File: aten/src/ATen/xpu/XPUEvent.h [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+
+  operator sycl::event&() const {
+    return event();
+  }
+
+  optional<at::Device> device() const {
+    if (isCreated()) {
+      return at::Device(at::kXPU, device_index_);
+    } else {
+      return {};
```
Comment (malfet) [inline aten/src/ATen/xpu/XPUEvent.h:48]:
```suggestion
      return nullopt;
```

File: aten/src/ATen/xpu/XPUEvent.h [hunk 23]
```diff
@@ -0,0 +1,145 @@
+#pragma once
+#include <ATen/xpu/XPUContext.h>
+#include <c10/util/Optional.h>
```
Comment (malfet) [inline aten/src/ATen/xpu/XPUEvent.h:3]: We are in C++ 17 land, std optional is there for everyone to use
```suggestion
#include <optional>
```

**Review by malfet — APPROVED** [review]

File: torch/csrc/xpu/Event.cpp (hunk 16 above)
Comment (guangyey) [reply torch/csrc/xpu/Event.cpp:113]: There are several reasons as follows:
1. for `torch.xpu.Stream`, we can use `sycl_queue` to get the raw pointer for sycl queue, and `torch.xpu.Stream.sycl_queue` is already used in `triton`. We would like to maintain naming consistency.
2. `sycl_event` precisely expresses that this is a SYCL event data type.
3. when the users migrate their code from cuda to sycl, naming `sycl_event` is easier and more convenient to modify from `cuda_event`.

Comment (guangyey) [reply torch/csrc/xpu/Event.cpp:113]: This is just my humble opinion. If CUDA also uses the name `raw_event_ptr`, it would be better for XPU to use it too.

File: aten/src/ATen/xpu/XPUEvent.h (hunk 20 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUEvent.h:53]: This will raise this error: 
```CPP
error: cannot convert ‘const std::unique_ptr<sycl::_V1::event>’ to ‘bool’ in return
```
std::unique_ptr is not implicitly convertible to bool. It is contextually convertible to bool (due to its [explicit conversion operator](http://en.cppreference.com/w/cpp/memory/unique_ptr/operator_bool)), which is why you can use it in an if statement, or put a ! in front of it, but you cannot pass it as an argument to a function which is expecting a bool.

File: aten/src/ATen/xpu/XPUEvent.h (hunk 21 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUEvent.h:52]: done.

File: aten/src/ATen/xpu/XPUEvent.h (hunk 18 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUEvent.h:22]: done. one with a default argument is better.

File: aten/src/ATen/xpu/XPUEvent.h (hunk 17 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUEvent.h:24]: Done.

File: aten/src/ATen/xpu/XPUEvent.h (hunk 23 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUEvent.h:3]: Done.

File: aten/src/ATen/xpu/XPUEvent.h (hunk 19 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUEvent.h]: I go through the code again, and indeed, this is not necessary. Removed.
